### PR TITLE
Add AdaBound (and AMSBound)

### PR DESCRIPTION
**Pull request by hitsgub**
This PR adds [AdaBound (ICLR2019)](https://openreview.net/forum?id=Bkg3g2R9FX).
This PR is the same as #6367, but the learning progress is confirmed in the local environment.
Note: The parameter `final_lr_rate` is different formation from original paper for exclusion of redundant calculation,
`final_lr_rate = alpha_star / alpha`.

### CODE REVIEW

File: chainer/optimizers/adam.py [hunk 1]
```diff
@@ -210,6 +277,18 @@ def lr(self):
...
         return self.alpha_t
 
+    @property
+    def bounds(self):
+        if self.t == 0:
+            raise RuntimeError(
+                'Can\'t determine the bounds of AdaBound optimizer '
+                'because the update steps have not been started.')
+        hp = self.hyperparam
+        final_lr = hp.final_lr_rate * hp.alpha
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: Did you test this with learning rate decay extensions? (I'm concerned about that this might be different from the original implementation)

**Comment by crcrpar** [comment]
Could you share some results?
For example, a plot of loss/accuracy that shows AdaBound's superiority.

This would also be one example: https://github.com/chainer/chainer/pull/4050#issuecomment-349835272

**Comment by hitsgub** [comment]
> Could you share some results?
> For example, a plot of loss/accuracy that shows AdaBound's superiority.
> 
> This would also be one example: [#4050 (comment)](https://github.com/chainer/chainer/pull/4050#issuecomment-349835272)

OK, I have just started training (ResNet18, [2-2-2-2]) 300 epoch, please wait for this evening.

File: chainer/optimizers/adam.py [hunk 2]
```diff
@@ -78,14 +89,20 @@ class AdamRule(optimizer.UpdateRule):
         eta (float): Schedule multiplier, can be used for warm restarts.
         weight_decay_rate (float): Weight decay rate.
         amsgrad (bool): Whether to use the AMSGrad variant of Adam.
+        adabound (bool): Whether to use the AdaBound variant of Adam.
+        final_lr (float): final (SGD) learning rate in adabound.
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: nit-picking: How about beginning the explanation of `final_lr` with a capital letter for consistency with the other arguments?
```suggestion
        final_lr (float): Final (SGD) learning rate in AdaBound.
```

File: chainer/optimizers/adam.py [hunk 3]
```diff
@@ -78,14 +89,20 @@ class AdamRule(optimizer.UpdateRule):
...
+        gamma (float): convergence speed of the bound functions in adabound.
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: Ditto
```suggestion
        gamma (float): Convergence speed of the bound functions in AdaBound.
```

File: chainer/optimizers/adam.py [hunk 4]
```diff
@@ -243,6 +327,9 @@ class Adam(optimizer.GradientMethod):
         eta (float): Schedule multiplier, can be used for warm restarts.
         weight_decay_rate (float): Weight decay rate.
         amsgrad (bool): Whether to use AMSGrad variant of Adam.
+        adabound (bool): Whether to use the AdaBound variant of Adam.
+        final_lr (float): final (SGD) learning rate in adabound.
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: nit-picking: Same as `AdamRule` (L93). 
How about using a capital letter for consistency with the other arguments?
```suggestion
        final_lr (float): Final (SGD) learning rate in AdaBound.
```

File: chainer/optimizers/adam.py [hunk 5]
```diff
@@ -243,6 +327,9 @@ class Adam(optimizer.GradientMethod):
...
+        gamma (float): convergence speed of the bound functions in adabound.
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: Ditto. (L94).
```suggestion
        gamma (float): Convergence speed of the bound functions in AdaBound.
```

**Comment by hitsgub** [comment]
Experimental result. (Some differences from original paper.)
- ResNet18(2-2-2-2) with pre-activation (original: ResNet34(3-4-6-3) with post-activation)
- Until 300epoch with reducing learning rate at [150, 225]. (original: 200epoch with reducing at [150])

test code repository: https://github.com/hitsgub/extra-chainer
 `python trainer.py --model models\PR2-2-2-2.py --opt [NesterovAG/Adam] (--amsgrad) (--adabound) --lr [0.1(NesterovAG)/0.001(others)]`

adabound

Like they say in original paper, AdaBound/AMSBound converge as fast as Adam/AMSGrad and as same accuracy as NesterovAG (like SGDM optimizer).
Especially in earlier epoch (200 epoch, used in original paper), AdaBound/AMSBound beats NesterovAG.

**Comment by crcrpar** [comment]
Hooray!

I really appreciate your work and the plot looks great.

File: chainer/optimizers/adam.py [hunk 6]
```diff
@@ -164,7 +194,45 @@ def update_core_gpu(self, param):
...
+            if AdamRule._amsbound_kernel is None:
+                AdamRule._amsbound_kernel = cuda.elementwise(
+                    'T grad, T alpha_t, T one_minus_beta1, T one_minus_beta2, '
+                    'T lower, T upper, '
+                    'T eps, T eta, T weight_decay_rate',
+                    'T param, T m, T v, T vhat',
+                    '''m += one_minus_beta1 * (grad - m);
+                       v += one_minus_beta2 * (grad * grad - v);
+                       vhat = max(vhat, v);
+                       param -= eta * (m * max(lower, min(upper, alpha_t / (sqrt(vhat) + eps))) +
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: Could you refactor this line to pass Python static check?
I guess one reference is L244-245 of this file.

> /home/travis/build/chainer/chainer/chainer/optimizers/adam.py:209:80: E501 line too long (97 > 79 characters)

File: chainer/optimizers/adam.py [hunk 7]
```diff
@@ -164,7 +194,45 @@ def update_core_gpu(self, param):
...
+        elif hp.adabound:
+            if AdamRule._adabound_kernel is None:
+                AdamRule._adabound_kernel = cuda.elementwise(
+                    'T grad, T alpha_t, T one_minus_beta1, T one_minus_beta2, '
+                    'T lower, T upper, '
+                    'T eps, T eta, T weight_decay_rate',
+                    'T param, T m, T v',
+                    '''m += one_minus_beta1 * (grad - m);
+                       v += one_minus_beta2 * (grad * grad - v);
+                       param -= eta * (m * max(lower, min(upper, alpha_t / (sqrt(v) + eps))) +
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: ditto.

> /home/travis/build/chainer/chainer/chainer/optimizers/adam.py:209:80: E501 line too long (97 > 79 characters)

File: chainer/optimizers/adam.py [hunk 8]
```diff
@@ -102,8 +119,15 @@ def __init__(self, parent_hyperparam=None,
...
             self.hyperparam.amsgrad = amsgrad
+        if adabound is not None:
+            self.hyperparam.adabound = adabound
+        if final_lr is not None:
+            self.hyperparam.final_lr = final_lr
+        if gamma is not None:
+            self.hyperparam.gamma = gamma
 
     def init_state(self, param):
+        self.state['initial_alpha'] = self.hyperparam.alpha
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: As of now, since only AdaBound uses `self.state['initial_alpha']`, so how about set this only when self.hyperparam.alpha is True?

e.g.
```python
if self.hyperparam.adabound:
    self.state['initial_alpha'] = self.hyperparam.alpha
```

File: chainer/optimizers/adam.py [hunk 9]
```diff
@@ -210,6 +278,18 @@ def lr(self):
...
         return self.alpha_t
 
+    @property
+    def bounds(self):
+        if self.t == 0:
+            raise RuntimeError(
+                'Can\'t determine the bounds of AdaBound optimizer '
+                'because the update steps have not been started.')
+        hp = self.hyperparam
+        final_lr = hp.final_lr * hp.alpha / self.state['initial_alpha']
```
Comment (crcrpar) [inline chainer/optimizers/adam.py]: Could you add a comment explaining that `hp.alpha / self.state['initial_alpha']` is a workaround to enable some of `chaienr.training.extensions` that change their target optimizer's parameters during training?

**Comment by crcrpar** [comment]
@beam2d 
The implementation and results seem good.
So, could you check this, especially the style and docs?

**Comment by beam2d** [comment]
Done a quick review. Looks good. Could you add parameters to tests? Just trying different values for `adabound` and `amsgrad` flags is ok (it looks the tests of adam+amsgrad is also lacking, so it would be good to add that too).

**Comment by hitsgub** [comment]
Thanks! Added arguments to testing.product.

**Comment by beam2d** [comment]
LGTM. Could you resolve the conflict?

**Comment by crcrpar** [comment]
Why are 116 files changed?

**Comment by hitsgub** [comment]
> Why are 116 files changed?

I googled a method to resolve conflict and I use `git rebase`, then all commits have merged into this PR.
Sorry.

**Comment by crcrpar** [comment]
The situation seems similar to #6253. 
Could you try the steps mentioned in [the comment](https://github.com/chainer/chainer/pull/6253#issuecomment-472261952)?

**Comment by crcrpar** [comment]
It seems fixed correctly and kudos to your quick work.

**Comment by crcrpar** [comment]
Hmm, this might conflict with #4365.

**Comment by beam2d** [comment]
I think it's good to first merge this PR for v6rc, and let #5217 adapt to it.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
It looks Adam tests are failing. Could you check them? (you can see the Jenkins log)

**Comment by hitsgub** [comment]
`self.state['xxx']` should be dynamic ndarray states, but there was `self.state['initial_alpha']` that is static float object.
fixed it to `self.initial.alpha`.

**Comment by hitsgub** [comment]
This CI error about `BatchRenormalization` seems to be unrelated to my PR.

**Comment by crcrpar** [comment]
@beam2d Could you take a look?

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by crcrpar** [comment]
@hitsgub could you check the failures?
They seem to be related to amsgrad.

**Comment by hitsgub** [comment]
@crcrpar OK! 
Amsgrad implementation is originally wrong and seems to be able to find it in my test addition.
https://github.com/chainer/chainer/pull/6388#issuecomment-470387281

**Comment by hitsgub** [comment]
I found amsgrad implementation is originally wrong as the lack of iDeep initialization.
I fixed it.

**Comment by hitsgub** [comment]
Raised errors are in `op_utils` and `GroupNormalization` now.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by hitsgub** [comment]
Because my using OS is Windows and cannot install iDeep, I am not sure if this fix works properly...
Currently, all bugs related to this PR are in `vhat = numpy.maximum(vhat, v, out=vhat)` with iDeep.

**Comment by crcrpar** [comment]
> Because my using OS is Windows and cannot install iDeep, I am not sure if this fix works properly...
> Currently, all errors related to this PR are in `numpy.maximum(vhat, v, out=vhat)` with iDeep.

Isn't Docker available?

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by hitsgub** [comment]
@crcrpar Now I started to learn about docker.

On the other hands, Jenkins test seems to be finished correctly (in spite of chainer-ci bot said 'Jenkins CI test filed.' in this thread).

**Comment by beam2d** [comment]
The CI failure looks irrelevant to this change. LGTM

**Comment by beam2d** [comment]
Thanks a lot!

**Comment by beam2d** [comment]
I will backport it to v6.

**Comment by crcrpar** [comment]
Kudos.
